catalog/seqexpr: index out of range [2] with length 2 during GetSequenceFromFunc while building function

Opening a new issue for the sqlsmith failure from https://github.com/cockroachdb/cockroach/issues/103258#issuecomment-1549031927:

---

roachtest.sqlsmith/setup=empty/setting=no-mutations [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/10104121?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/10104121?buildTab=artifacts#/sqlsmith/setup=empty/setting=no-mutations) on master @ [107ad579b1ff8311ac4f4df4783985010a0322f9](https://github.com/cockroachdb/cockroach/commits/107ad579b1ff8311ac4f4df4783985010a0322f9):


```
test artifacts and logs in: /artifacts/sqlsmith/setup=empty/setting=no-mutations/run_1
(sqlsmith.go:248).func3: error: pq: internal error: runtime error: index out of range [2] with length 2
stmt:
CREATE FUNCTION func😒120(IN p0 GEOMETRY, IN p1 BOX2D, IN p2 BOX2D, IN p3 BYTES, IN p4 REGCLASS, IN p5 CHAR)
	RETURNS SETOF TSVECTOR
	RETURNS NULL ON NULL INPUT
	VOLATILE
	NOT LEAKPROOF
	LANGUAGE SQL
	AS $funcbody$SELECT '''BaLWsnOlj'':56C,585C ''GqMInY'':740A,775A ''Ni'':268B,275B,398B,593C ''Prdh'':886A,929C ''RXbtNU'' ''SwVjDEwHQU'' ''XblZ'' ''akkxxS'':234C,317A,832C ''dnAzp'':504B,767B ''g''':::TSVECTOR AS "coL10495" FROM (VALUES ((1310369687:::OID,)), ((SELECT (1536774293:::OID,) AS " coL10493" FROM (VALUES (p4), (NULL), (p4), (3469644366:::OID), (p4)) AS tab5685 (col10486), (VALUES (p4), (p4), (p4), (1114601105:::OID)) AS "t!ab5686" ("cOl10487"), (VALUES (ARRAY[(-28177):::INT8], ('170.240.174.81/11':::INET::INET - (-6649095639022782424):::INT8::INT8)::INET), (ARRAY[(-25104):::INT8], '93.47.227.69/6':::INET), (ARRAY[22170:::INT8,(-1):::INT8,(-2279):::INT8], 'ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff':::INET), (ARRAY[(-21234):::INT8,11613:::INT8], (('498e:4494:e0c9:9ee7:a2b5:84b2:d4af:1be/66':::INET::INET | '93.238.93.91/7':::INET::INET)::INET::INET + CASE WHEN true THEN setval(p4::REGCLASS, ('229.224.8.237':::INET::INET - NULL::INET)::INT8::INT8)::INT8 ELSE 515316040633998673:::INT8 END::INT8)::INET), (NULL, set_masklen('47.35.96.191/30':::INET::INET, array_position(ARRAY['2025-01-03 04:08:53.000791':::TIMESTAMP,'2009-12-09 23:06:17.000347':::TIMESTAMP,'2015-07-22 01:23:56.000865':::TIMESTAMP]::TIMESTAMP[], '1994-10-03 03:32:33.000732':::TIMESTAMP::TIMESTAMP)::INT8::INT8)::INET), ((SELECT ARRAY[24712:::INT8,0:::INT8,(-32252):::INT8,(-365):::INT8] AS col10490 FROM (VALUES ('6825e8a8-ce4a-4b9b-879f-292833e4d004':::UUID), ('97ce63fb-8d4c-454d-ab31-17acda8d2ab5':::UUID), ('d3fadccf-62fa-4b9e-ae95-7c046bb2154b':::UUID)) AS tab5687 (col10488), (VALUES ('1977-01-05 08:08:53.00024':::TIMESTAMP), (parse_timestamp(e'\t.\x14I=':::STRING::STRING)::TIMESTAMP), ('1986-04-09 04:01:35.000584':::TIMESTAMP)) AS " 😀ta̠b\\U00032BC45688" (" c""ol10489") ORDER BY " 😀ta̠b\\U00032BC45688"." c""ol10489" NULLS LAST, " 😀ta̠b\\U00032BC45688"." c""ol10489" ASC NULLS LAST, " 😀ta̠b\\U00032BC45688"." c""ol10489" DESC NULLS FIRST LIMIT 1:::INT8), NULL)) AS tab5689 ("c'ol10491", col10492) LIMIT 1:::INT8))) AS tab5690 (col̫10494) ORDER BY tab5690.col̫10494 DESC NULLS FIRST, tab5690.col̫10494 NULLS LAST LIMIT 28:::INT8$funcbody$;
```
<p>Parameters: <code>ROACHTEST_cloud=gce</code>
, <code>ROACHTEST_cpu=4</code>
, <code>ROACHTEST_encrypted=false</code>
, <code>ROACHTEST_ssd=0</code>
</p>
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*sqlsmith/setup=empty/setting=no-mutations.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-28233

This tripped up another, [unrelated CI run](https://teamcity.cockroachdb.com/viewLog.html?tab=buildLog&logTab=tree&filter=debug&expand=all&buildId=10267972).

```
	test artifacts and logs in: /artifacts/sqlsmith/setup=empty/setting=no-mutations/run_1
(sqlsmith.go:240).func3: error: pq: internal error: 35 trailing bytes in encoded value: [15 15 20 64 0 0 1 144 0 0 4 144 0 0 8 116 101 115 116 106 115 111 110 15 16 0 8 148 245 167 155 5 176 156 27]
stmt:
WITH
	with5045 ("c%qol41326")
		AS (
			SELECT
				*
			FROM
				(
					VALUES
						(🙄func273()),
						((NULL, '{"test": "json"}':::JSONB, (), '1992-03-04 16:39:38.000223+00':::TIMESTAMPTZ)),
						(
							(NULL, '[{"EvG": [true, "baz", false], "MPRRONv": 0.2980103397565371, "foo": {"8rNz5y": true, "baz": null, "kRMYs": false}, "uyRcxposI": "foo"}, {"1rMfbBSOg": {"USyO": true}, "Li5": 1.261090300252655, "a": "7c7QH82", "nsAleYB": "rga3g"}, {"bar": 2.439379404706056}, true, 0.23533439446032228]':::JSONB, (), NULL)
						)
				)
					AS "\\uE1B1t%Ab 21645" ("c%qol41326")
		)
SELECT
	"c%40te_ref1478"."c%qol41326" AS "c͚oL41327"
FROM
	with5045 AS "c%40te_ref1478"
GROUP BY
	"c%40te_ref1478"."c%qol41326"
HAVING
	bool_and(true::BOOL)::BOOL
LIMIT
	95:::INT8;
[20:17:12]	--- FAIL: sqlsmith/setup=empty/setting=no-mutations (413.48s)
```

--- a/pkg/sql/catalog/seqexpr/sequence.go
+++ b/pkg/sql/catalog/seqexpr/sequence.go
@@ -80,21 +80,20 @@ func GetSequenceFromFunc(funcExpr *tree.FuncExpr) (*SeqIdentifier, error) {
 
 	if hasSequenceArguments {
 		found := false
-		for i := range def.Overloads {
+		for _, overload := range def.Overloads {
 			// Find the overload that matches funcExpr.
-			if len(funcExpr.Exprs) == def.Overloads[i].Types.Length() {
-				found = true
-				paramTypes, ok := def.Overloads[i].Types.(tree.ParamTypes)
+			if len(funcExpr.Exprs) == overload.Types.Length() {
+				paramTypes, ok := overload.Types.(tree.ParamTypes)
 				if !ok {
 					panic(pgerror.Newf(
 						pgcode.InvalidFunctionDefinition,
 						"%s has invalid argument types", funcExpr.Func.String(),
 					))
 				}
-				for i := 0; i < def.Overloads[i].Types.Length(); i++ {
+				found = true
+				for i := 0; i < len(paramTypes); i++ {
 					// Find the sequence name param.
-					paramName := paramTypes[i].Name
-					if paramName == builtinconstants.SequenceNameArg {
+					if paramTypes[i].Name == builtinconstants.SequenceNameArg {
 						arg := funcExpr.Exprs[i]
 						if seqIdentifier := getSequenceIdentifier(arg); seqIdentifier != nil {
 							return seqIdentifier, nil

--- a/pkg/sql/logictest/testdata/logic_test/udf
+++ b/pkg/sql/logictest/testdata/logic_test/udf
@@ -3475,3 +3475,32 @@ SELECT f100915(20-(SELECT y FROM (VALUES (10), (20)) y(y) WHERE x=y)) FROM (VALU
 ----
 true
 true
+
+subtest regression_103869
+
+statement ok
+CREATE SEQUENCE sq_103869;
+
+statement ok
+CREATE FUNCTION f_103869(sq REGCLASS) RETURNS INT
+LANGUAGE SQL
+AS $$
+    SELECT setval(sq, 1);
+$$;
+
+query I
+SELECT f_103869('sq_103869'::REGCLASS);
+----
+1
+
+statement ok
+CREATE FUNCTION f_103869(sq STRING) RETURNS INT
+LANGUAGE SQL
+AS $$
+    SELECT setval(sq, 2);
+$$
+
+query I
+SELECT f_103869('sq_103869')
+----
+2

